src/services/jsTyping.ts

```diff
@@ -0,0 +1,288 @@
+                getTypingNamesFromJson(bowerJsonPath, filesToWatch);
+
+                const nodeModulesPath = ts.combinePaths(searchDir, "node_modules");
+                getTypingNamesFromNodeModuleFolder(nodeModulesPath, filesToWatch);
+            }
+
+            getTypingNamesFromSourceFileNames(fileNames);
+            getTypingNamesFromCompilerOptions(compilerOptions);
+        }
+
+        const typingsPath = ts.combinePaths(cachePath, "typings");
+        const tsdJsonPath = ts.combinePaths(cachePath, "tsd.json");
+        const tsdJsonDict = tryParseJson(tsdJsonPath, host);
+        if (tsdJsonDict) {
+            for (const notFoundTypingName of notFoundTypingNames) {
+                if (inferredTypings.hasOwnProperty(notFoundTypingName) && !inferredTypings[notFoundTypingName]) {
+                    delete inferredTypings[notFoundTypingName];
+                }
+            }
+
+            // The "installed" property in the tsd.json serves as a registry of installed typings. Each item
+            // of this object has a key of the relative file path, and a value that contains the corresponding
+            // commit hash.
+            if (hasProperty(tsdJsonDict, "installed")) {
+                for (const cachedTypingPath in tsdJsonDict.installed) {
+                    // Assuming the cachedTypingPath has the format of "[package name]/[file name]"
+                    const cachedTypingName = cachedTypingPath.substr(0, cachedTypingPath.indexOf("/"));
+                    // If the inferred[cachedTypingName] is already not null, which means we found a corresponding
+                    // d.ts file that coming with the package. That one should take higher priority.
+                    if (hasProperty(inferredTypings, cachedTypingName) && !inferredTypings[cachedTypingName]) {
+                        inferredTypings[cachedTypingName] = ts.combinePaths(typingsPath, cachedTypingPath);
+                    }
+                }
+            }
+        }
+
+        // Remove typings that the user has added to the exclude list
+        for (const excludeTypingName of exclude) {
+            delete inferredTypings[excludeTypingName];
+        }
+
+        const newTypingNames: string[] = [];
+        const cachedTypingPaths: string[] = [];
+        for (const typing in inferredTypings) {
+            if (inferredTypings[typing] !== undefined) {
+                cachedTypingPaths.push(inferredTypings[typing]);
+            }
+            else {
+                newTypingNames.push(typing);
+            }
+        }
+        return { cachedTypingPaths, newTypingNames, filesToWatch };
+
+        /**
+         * Merge a given list of typingNames to the inferredTypings map
+         */
+        function mergeTypings(typingNames: string[]) {
+            if (!typingNames) {
+                return;
+            }
+
+            for (const typing of typingNames) {
+                if (!inferredTypings.hasOwnProperty(typing)) {

```

Comment:
In cases when a node package comes with its own .d.ts files, the `inferredTypings[typing]` may have a actual value, which is the .d.ts file name. So the `if` check is still needed as we don't want to override that value.